File: benchmark/routes/render_dedup/mixed_region_feed_ivar.rb
# frozen_string_literal: true

# Ivar-shaped mixed-region render-dedup benchmark.
#
# Drives `benchmark/k6/render_dedup/mixed_region_feed_ivar.js`. Mirrors
# the structure of the sibling `mixed_region_feed` runner but:
#
# - The benchmark app's `/featured_item` page renders a partial whose
#   dynamic sources are controller ivars (`@featured_item.title`,
#   `@featured_item.body`).
# - At subscribe time, `SlotStateCapture` resolves each ivar via
#   `view_assigns` threaded through `FragmentRegistrationMetadata::Builder`,
#   populating `fragment_slot_states["value"]` for the ivar-bound slots.
# - One writer updates the same FeedItem row; every subscriber should
#   receive a byte-equality-proven patch (no synthetic-request render)
#   because the proof gate now sees a populated binding.
#
# Pass criteria mirror the parent `mixed_region_feed` set: zero render
# Render call errors, zero classification downgrades, and at least one proof
# activity signal — either render-path region outcomes (the older
# region-grouping path) or `:byte_equality` proven verdicts (the
# direct-push proof path that ships ops without rendering). The
# parent runner now exposes `byte_equality_proofs` in the report so
# subclasses inherit proof-aware acceptance without overriding
# `status_for`.

require_relative "mixed_region_feed"

module Upkeep
  module Benchmark
    class MixedRegionFeedIvar < MixedRegionFeed
      private

      def workload_script
        File.expand_path("../../k6/render_dedup/mixed_region_feed_ivar.js", __dir__)
      end

      # In addition to the parent's "any proof activity" check, this
      # workload's defining property is that every invalidation must
      # prove byte-equality: the page renders a single ivar-bound
      # fragment whose only dynamic source is `@featured_item.title`,
      # so any non-proven verdict is a regression in the ivar code
      # path. Keep the parent assertions, then
      # add a strict `byte_equality_proofs > 0` check on top.
      def status_for(report, dispatch_metrics, workload_status)
        base_status = super
        return base_status unless base_status.zero?

        if report[:byte_equality_proofs].zero?
          warn("[mixed_region_feed_ivar] byte_equality proof verdicts = 0 — ivar proof path did not fire")
          return 2
        end

        0
      end

      def write_report(report)
        out = File.join(results_dir, "render-dedup-mixed-region-feed-ivar-#{timestamp}.json")
        File.write(out, JSON.pretty_generate(report))
      end

      def print_report(report)
        super
        puts "  (workload variant: ivar; byte-equality proof verdicts are required)"
      end
    end
  end
end

if $PROGRAM_NAME == __FILE__
  opts = {
    base_url: ENV["BENCH_BASE_URL"] || "http://localhost:3000",
    ws_url: ENV["BENCH_WS_URL"] || "ws://localhost:3000/cable",
    metrics_url: ENV["RELAY_METRICS_URL"] || "http://127.0.0.1:9394/metrics",
    vus: Integer(ENV["BENCH_VUS"] || "50"),
    results_dir: ENV["BENCH_RESULTS_DIR"] || File.expand_path("../../results", __dir__),
    timestamp: ENV["BENCH_TIMESTAMP"] || Time.now.strftime("%Y%m%d%H%M%S")
  }

  OptionParser.new do |o|
    o.on("--base-url=URL") { |v| opts[:base_url] = v }
    o.on("--ws-url=URL") { |v| opts[:ws_url] = v }
    o.on("--metrics-url=URL") { |v| opts[:metrics_url] = v }
    o.on("--vus=N", Integer) { |v| opts[:vus] = v }
    o.on("--results-dir=DIR") { |v| opts[:results_dir] = v }
    o.on("--timestamp=TS") { |v| opts[:timestamp] = v }
  end.parse!

  exit Upkeep::Benchmark::MixedRegionFeedIvar.new(**opts).run
end

File: benchmark/k6/render_dedup/mixed_region_feed_ivar.js
// Ivar-shaped mixed-region render-dedup scenario.
//
// N authenticated VUs subscribe to the singular FeaturedItems resource.
// The page renders a fragment whose dynamic sources are controller
// ivars (`@featured_item.title`, `@featured_item.body`). At subscribe
// time, `SlotStateCapture` resolves each ivar via `view_assigns`
// threaded through `FragmentRegistrationMetadata::Builder`, populating
// `fragment_slot_states["value"]` for the ivar-bound slots. One writer
// updates the featured item; every subscriber should receive a
// byte-equality-proven patch (no synthetic-request render) because the
// proof gate now sees a populated binding.
//
// This workload is the F3 follow-up that exercises the new ivar code
// path. The sibling `mixed_region_feed.js` workload uses local-receiver
// templates that were already covered before the
// `direct-push-proof-expansion` candidate; this workload is the
// counterpart that drives the new view-assigns-derived bindings.

import http from "k6/http";
import ws from "k6/ws";
import { check, fail } from "k6";
import { Counter, Trend } from "k6/metrics";
import { BASE_URL, WS_URL, NUM_USERS } from "../utils/config.js";
import { login } from "../utils/auth.js";
import { findBetween } from "../utils/index.js";
import { textSummary } from "../utils/summary.js";
import { relayWsUrl } from "../utils/relay_scenario.js";

const rtt = new Trend("rtt", true);
const writesIssued = new Counter("writes_issued");
const deliveries = new Counter("deliveries_observed");

const N_VUS = parseInt(__ENV.BENCH_VUS || "50", 10);
const STEADY_SECONDS = parseInt(__ENV.IVAR_FEED_STEADY_S || "20", 10);
const N_WRITERS = parseInt(__ENV.FEATURED_WRITERS || "1", 10);

export const options = {
  summaryTrendStats: ["avg", "min", "med", "max", "p(50)", "p(90)", "p(95)", "p(99)"],
  thresholds: {
    checks: ["rate==1"],
    writes_issued: [`count>=${N_WRITERS}`],
    deliveries_observed: [`count>=${N_VUS - N_WRITERS}`],
  },
  scenarios: {
    mixed_region_feed_ivar: {
      executor: "per-vu-iterations",
      vus: N_VUS,
      iterations: 1,
      maxDuration: `${STEADY_SECONDS + 30}s`,
    },
  },
};

export default function () {
  const userIdx = ((__VU - 1) % NUM_USERS) + 1;
  const email = `user${userIdx}@bench.test`;
  const jar = login(BASE_URL, email, "benchpass123");

  const pageRes = http.get(`${BASE_URL}/featured_item`, { jar });
  check(pageRes, { "featured_item loaded": (r) => r.status === 200 });

  const token = findBetween(pageRes.body, 'data-context-token="', '"');
  if (!token) fail("could not extract context_token from /featured_item");

  const wsResult = ws.connect(relayWsUrl(WS_URL, token), {}, function (sock) {
    let sendTs = null;

    sock.on("binaryMessage", function () {
      deliveries.add(1);
      if (sendTs !== null) {
        rtt.add(Date.now() - sendTs);
        sendTs = null;
      }
    });
    sock.setInterval(function () {}, 1000);

    if (__VU <= N_WRITERS) {
      const offsetMs = N_WRITERS > 1 ? ((__VU - 1) * 1000) / N_WRITERS : 0;
      sock.setTimeout(function () {
        sendTs = Date.now();
        const nonce = `${__VU}-${sendTs}`;
        const res = http.patch(
          `${BASE_URL}/featured_item`,
          { title: `Featured title ${nonce}`, body: `Featured body ${nonce}` },
          { jar, tags: { name: "PATCH_featured_item" } }
        );
        check(res, { "featured update 2xx": (r) => r.status >= 200 && r.status < 300 });
        writesIssued.add(1);
      }, 3000 + offsetMs);
    }

    sock.setTimeout(function () { sock.close(); }, STEADY_SECONDS * 1000);
  });

  check(wsResult, { "ws connected": (r) => r && r.status === 101 });
}

export function handleSummary(data) {
  return {
    stdout: textSummary(data, { indent: " ", enableColors: true }),
    "results/render-dedup-mixed-region-feed-ivar.json": JSON.stringify(data),
  };
}
